main/webapp/modules/core/scripts/views/data-table/menu-edit-column.js

```diff
@@ -320,14 +334,204 @@ DataTableColumnHeaderUI.extendMenu(function(column, columnHeaderUI, menu) {
       );
       dismiss();
     });
+  }; 
+  
+  var doJoinColumns = function() {
+    var self = this;
+    self.onError = "keep-original" ;
+    self.repeat = false ;
+    self.repeatCount = "";
+    /*
+     * Close the dialog window
+     */
+    var dismiss = function() {
+       DialogSystem.dismissUntil(self.level - 1);
+     };
+     /*
+     * Join the columns according to user input
+     */
+    var transform = function() {
+      // get options
+      self.deleteJoinedColumns = self.elmts.delete_joined_columnsInput[0].checked;
+      self.writeOrCopy = $("input[name='write-or-copy']:checked")[0].value;
+      self.newColumnName = self.elmts.new_column_nameInput[0].value;
+      self.manageNulls = $("input[name='manage-nulls']:checked")[0].value;
+      self.nullSubstitute = self.elmts.null_substituteInput[0].value;
+      self.fieldSeparator = self.elmts.field_separatorInput[0].value;
+      // correct options if they are not consistent
+      if (self.newColumnName != "") {
+        self.writeOrCopy ="copy-to-new-column";
+        } else
+        {
+          self.writeOrCopy ="write-active-column";
+        };
+      if (self.nullSubstitute != "") {
+          self.manageNulls ="replace-nulls";
+      };
+      // build GREL expression
+      var columnsToJoin = [];
+      self.elmts.column_join_columnPicker
+        .find('.column-join-column input[type="checkbox"]:checked')
+        .each(function() {
+            columnsToJoin.push (this.closest ('.column-join-column').getAttribute('column'));
+         });
+      self.expression = columnsToJoin.map (function (colName) {
+        if (self.manageNulls == "skip-nulls") {
+          return "cells['"+colName+"'].value";
+        }
+        else {
+          return "coalesce(cells['"+colName+"'].value, '"+self.nullSubstitute+ "')";
+        }
+      }).join (',');
+      self.expression = 'join ([' + self.expression + '],"' + self.fieldSeparator + '")';
+      // apply expression to active column or new column
+      if (self.writeOrCopy =="copy-to-new-column") {
+        Refine.postCoreProcess(
+          "add-column", 
+          {
+          baseColumnName: column.name,  
+          newColumnName: self.newColumnName, 
+          columnInsertIndex: columnIndex + 1,
+          onError: self.onError
+          },
+          { expression: self.expression },
+          { modelsChanged: true }
+        );
+      } 
+      else {
+        doTextTransform(column.name, self.expression, self.onError, self.repeat, self.repeatCount);
+      }
+      // delete joined columns
+    if (self.deleteJoinedColumns) {
+      // do not delete the active cell if it contains the result
+      if (self.writeOrCopy !="copy-to-new-column") {
+        columnsToJoin = columnsToJoin.filter(function(item) {
+          return item !== column.name
+      });
+      }
+      columnsToJoin.forEach (function (columnName) {
```

Comment:
How about using just one operation to remove all the necessary columns at once? The `column-reorder` operation can do this.

About the nature of the error: due to the asynchronous nature of Javascript, making HTTP requests in a loop like here tends to be risky, as they will all be fired at the same time (so you do not control in which order they will reach the server). It is better to pass `postCoreProcess` a callback function which will continue making the next calls once the first one has completed. This also holds for the first call above (to `add-column`): you should wait for this HTTP request to complete before deleting the other columns. I suspect that is the reason why you are having issues here: some column removal operation might be executed before the join itself.

You can pass a callback to `Refine.postCoreProcess` as follows:

```javascript
var onJoinSuccess = function() {
    // Do other calls to Refine.postCoreProcess to remove columns here
}

var onJoinError = function(e) {
    // Report to the user that we could not join the columns for some reason (given in e)
}

Refine.postCoreProcess("add-column", parameters,
   { expression: self.expression },
   { modelsChanged: true },
   { onDone:onJoinSuccess, onError:onJoinError });
```

Reply:
Thanks! I did not realized that the HTTP requests between frontend and backend were managed asynchronously like an ordinary web request